File: 0004.py
# O(n)
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        nums1.extend(nums2)
        nums = sorted(nums1)
        if len(nums) % 2 == 0:
            return (nums[len(nums) // 2] + nums[len(nums) // 2 - 1]) / 2
        else:
            return nums[len(nums) // 2]


# O(log (m+n)):
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        l1 = len(nums1)
        l2 = len(nums2)
        # To make sure j > 0
        if l1 > l2:
            nums1, nums2 = nums2, nums1
            l1, l2 = l2, l1
        if l1 == 0:
            if l2 % 2 != 0:
                return nums2[l2 // 2]
            return (nums2[l2 // 2] + nums2[l2 // 2 - 1]) / 2
        # 'mid' is right median
        mid = (l1 + l2) // 2
        l = 0
        r = l1
        while l <= r:
            i = (l + r) // 2
            j = mid - i
            if i < l1 and nums1[i] < nums2[j - 1]:
                l = i + 1
            elif i > 0 and nums1[i - 1] > nums2[j]:
                r = i - 1
            else:
                # check left median
                if i == 0:
                    left_mid = nums2[j - 1]
                elif j == 0:
                    left_mid = nums1[i - 1]
                else:
                    left_mid = max(nums1[i - 1], nums2[j - 1])
                # check right median
                if i == l1:
                    right_mid = nums2[j]
                elif j == l2:
                    right_mid = nums1[i]
                else:
                    right_mid = min(nums1[i], nums2[j])

                if (l1 + l2) % 2 != 0:
                    return right_mid
                return (left_mid + right_mid) / 2

File: 0004.cpp
// First time
/*
    time complexity should be O(log (m+n)) -> Binary search!!

    Just consider the index
    nums1:    0 ... m1-1, m1 ... n1-1
                      ^   ^
                   left1  right1
    nums2:    0 ... m2-1, m2 ... n2-1
                      ^   ^
                   left2  right2
    merge(v): 0 ... k-1, k ... n1+n2-1 -> odd: v[k], even: (v[k]+v[k-1])/2
             |_________|
               k = m1 + m2
    v[k-1] = max(nums1[m1-1], nums[m2-1])
    v[k] = min(nums1[m1], nums2[m2])

    Use binary search to guess m1
    ex.
    odd:  0, (1), 2    -> 3/2 = 1 = k
    even: 0, (1, 2), 3 -> 4/2 = 2 = k

    T: O(log(m+n))/S: O(1)
*/
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() > nums2.size())
            swap(nums1, nums2);
        int n1 = nums1.size(), n2 = nums2.size();
        int l = 0, r = n1;
        while (l <= r) {
            int m1 = l + (r - l) / 2;
            int m2 = (n1 + n2) / 2 - m1;

            int left1 = m1 == 0 ? INT_MIN : nums1[m1 - 1];
            int right1 = m1 == n1 ? INT_MAX : nums1[m1];
            int left2 = m2 == 0 ? INT_MIN : nums2[m2 - 1];
            int right2 = m2 == n2 ? INT_MAX : nums2[m2];

            if (left1 <= right2 && left2 <= right1) {
                if ((n1 + n2)&1)
                    return min(right1, right2);
                else
                    return (max(left1, left2) + min(right1, right2)) / 2.0;
            }

            if (right1 < left2)
                l = m1 + 1;
            else
                r = m1 - 1;
        }
        return 0.0;
    }
};